services/src/datasets/external/sentinel_s2_l2a_cogs.rs

```diff
@@ -0,0 +1,640 @@
+use crate::stac::FeatureCollection as StacCollection;
+use crate::stac::StacAsset;
+use crate::{datasets::listing::DatasetListOptions, error::Result};
+use crate::{
+    datasets::{
+        listing::{DatasetListing, DatasetProvider},
+        storage::DatasetProviderDefinition,
+    },
+    error,
+    util::user_input::Validated,
+};
+use async_trait::async_trait;
+use chrono::{DateTime, Duration, Utc};
+use geoengine_datatypes::dataset::{DatasetId, DatasetProviderId, ExternalDatasetId};
+use geoengine_datatypes::primitives::{Measurement, TimeInterval};
+use geoengine_datatypes::raster::GeoTransform;
+use geoengine_datatypes::raster::RasterDataType;
+use geoengine_datatypes::spatial_reference::{SpatialReference, SpatialReferenceAuthority};
+use geoengine_operators::engine::QueryRectangle;
+use geoengine_operators::source::GdalDatasetParameters;
+use geoengine_operators::source::GdalLoadingInfoPart;
+use geoengine_operators::source::GdalLoadingInfoPartIterator;
+use geoengine_operators::{
+    engine::{MetaData, MetaDataProvider, RasterResultDescriptor, VectorResultDescriptor},
+    mock::MockDatasetDataSourceLoadingInfo,
+    source::{GdalLoadingInfo, OgrSourceDataset},
+};
+use log::debug;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::collections::HashMap;
+use std::path::PathBuf;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SentinelS2L2ACogsProviderDefinition {
+    name: String,
+    id: DatasetProviderId,
+    api_url: String,
+}
+
+#[typetag::serde]
+#[async_trait]
+impl DatasetProviderDefinition for SentinelS2L2ACogsProviderDefinition {
+    async fn initialize(
+        self: Box<Self>,
+    ) -> crate::error::Result<Box<dyn crate::datasets::listing::DatasetProvider>> {
+        Ok(Box::new(SentinelS2L2aCogsDataProvider::new(
+            self.id,
+            self.api_url,
+        )))
+    }
+
+    fn type_name(&self) -> String {
+        "SentinelS2L2ACogs".to_owned()
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn id(&self) -> DatasetProviderId {
+        self.id
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Band {
+    pub name: String,
+    pub no_data_value: Option<f64>,
+    pub data_type: RasterDataType,
+}
+
+impl Band {
+    pub fn new(name: String, no_data_value: Option<f64>, data_type: RasterDataType) -> Self {
+        Self {
+            name,
+            no_data_value,
+            data_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Zone {
+    pub name: String,
+    pub epsg: u32,
+}
+
+impl Zone {
+    pub fn new(name: String, epsg: u32) -> Self {
+        Self { name, epsg }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SentinelMetaData {
+    bands: Vec<Band>,
+    zones: Vec<Zone>,
+}
+
+#[derive(Debug, Clone)]
+pub struct SentinelDataset {
+    band: Band,
+    zone: Zone,
+    listing: DatasetListing,
+}
+
+pub struct SentinelS2L2aCogsDataProvider {
+    #[allow(dead_code)]
```

Comment:
why is this dead?

Reply:
it's dead right now because I create all the MetaData on `new` but in practice it should be retrieved from STAC and in order to generate the external dataset id the provider needs to know its own provider id, which is part of the dataset id. I can remove it for now if you want

Reply:
add a todo

Reply:
3ed69c0b2d4e186ec765d553ddc10d46fc780e39